src/Common/AbstractEnum.php

```diff
@@ -0,0 +1,356 @@
+     *
+     * @param string|int $value The enum value
+     * @return static
+     * @throws InvalidArgumentException If the value is not valid
+     * @deprecated Use from() method instead
+     */
+    public static function fromValue($value): self
+    {
+        return self::from($value);
+    }
+
+    /**
+     * Get or create a singleton instance for the given value and name
+     *
+     * @param string|int $value The enum value
+     * @param string $name The constant name
+     * @return static
+     * @phpstan-return static
+     */
+    private static function getInstance($value, string $name): self
+    {
+        $className = static::class;
+        $key = $name;
+
+        if (!isset(self::$instances[$className])) {
+            self::$instances[$className] = [];
+        }
+
+        if (!isset(self::$instances[$className][$key])) {
+            $instance = new $className($value, $name);
+            self::$instances[$className][$key] = $instance;
+        }
+
+        /** @var static */
+        return self::$instances[$className][$key];
+    }
+
+    /**
+     * Get all constants for this enum class
+     *
+     * @return array<string, string|int>
+     * @throws \RuntimeException If invalid constant found
+     */
+    protected static function getConstants(): array
+    {
+        $className = static::class;
+
+        if (!isset(self::$cache[$className])) {
+            $reflection = new ReflectionClass($className);
+            $constants = $reflection->getConstants();
+
+            // Validate all constants
+            $enumConstants = [];
+            foreach ($constants as $name => $value) {
+                // Check if constant name follows uppercase snake_case pattern
+                if (!preg_match('/^[A-Z][A-Z0-9_]*$/', $name)) {
+                    throw new \RuntimeException(
+                        sprintf(
+                            'Invalid enum constant name "%s" in %s. Constants must be UPPER_SNAKE_CASE.',
+                            $name,
+                            $className
+                        )
+                    );
+                }
+
+                // Check if value is valid type
+                if (!is_string($value) && !is_int($value)) {
+                    throw new \RuntimeException(
+                        sprintf(
+                            'Invalid enum value type for constant %s::%s. ' .
+                            'Only string and int values are allowed, %s given.',
+                            $className,
+                            $name,
+                            gettype($value)
+                        )
+                    );
+                }
+
+                $enumConstants[$name] = $value;
+            }
+
+            self::$cache[$className] = $enumConstants;
+        }
+
+        return self::$cache[$className];
+    }
+
+    /**
+     * Handle dynamic method calls for enum creation and checking
```

Comment:
"enum _creation_ and checking"? As far as I can tell, this only considers checking, not creating.